Fix SecretRef validation for PersistentVolumes

akshitgrover opened the pull request:
- Add `SecretRef` validation for core.RBDPersistentVolumeSource
- Add `SecretRef` validation for core.CephFSPersistentVolumeSource
- Add `Name` and `Namespace` value validation for core.ISCSIPersistentVolumeSource.SecretRef
- Add `Name` and `Namespace` value validation for core.CinderPersistentVolumeSource.SecretRef
- Add `SecretRef` validation for core.FlexPersistentVolumeSource
- Add `SecretName` and `SecretNamespace` validation for core.AzureFilePersistentVolumeSource
- Add `SecretRef` validation for core.ScaleIOPersistentVolumeSource
- Add `Name` and `Namespace` value validation for core.StorageOSPersistentVolumeSource.SecretRef
- Add `Name` and `Namespace` value validation for core.CSIPersistentVolumeSource.SecretRef

Signed-off-by: Akshit Grover <akshit.grover2016@gmail.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Adds and fixes validation checks for SecretReferences in PersistentVolumes

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #102939

#### Special notes for your reviewer:
Adds SecretReference checks for PersistentVolumes wherever missing

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

akgrover committed:
Fix SecretRef validation for PersistentVolumes

- Add `SecretRef` validation for core.RBDPersistentVolumeSource
- Add `SecretRef` validation for core.CephFSPersistentVolumeSource
- Add `Name` and `Namespace` value validation for core.ISCSIPersistentVolumeSource.SecretRef
- Add `Name` and `Namespace` value validation for core.CinderPersistentVolumeSource.SecretRef
- Add `SecretRef` validation for core.FlexPersistentVolumeSource
- Add `SecretName` and `SecretNamespace` validation for core.AzureFilePersistentVolumeSource
- Add `SecretRef` validation for core.ScaleIOPersistentVolumeSource
- Add `Name` and `Namespace` value validation for core.StorageOSPersistentVolumeSource.SecretRef
- Add `Name` and `Namespace` value validation for core.CSIPersistentVolumeSource.SecretRef

Signed-off-by: Akshit Grover <akshit.grover2016@gmail.com>

akshitgrover commented:
/sig storage

akshitgrover commented:
/cc @jsafrane

akshitgrover commented:
/assign @jsafrane

akshitgrover commented:
PTAL @jsafrane 
Thank you

jsafrane commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1262,6 +1271,21 @@ func validateRBDPersistentVolumeSource(rbd *core.RBDPersistentVolumeSource, fldP
 	if len(rbd.RBDImage) == 0 {
 		allErrs = append(allErrs, field.Required(fldPath.Child("image"), ""))
 	}
+	if rbd.SecretRef != nil {
+		if len(rbd.SecretRef.Name) == 0 {
+			allErrs = append(allErrs, field.Required(fldPath.Child("secretRef", "name"), ""))
+		} else {
+			for _, msg := range ValidateSecretName(rbd.SecretRef.Name, false) {
+				allErrs = append(allErrs, field.Invalid(fldPath.Child("secretRef", "name"), rbd.SecretRef.Name, msg))
+			}
+		}
+		if len(rbd.SecretRef.Namespace) == 0 {
+			allErrs = append(allErrs, field.Required(fldPath.Child("secretRef", "namespace"), ""))
+		} else {
+			allErrs = append(allErrs, ValidateDNS1123Label(rbd.SecretRef.Namespace, fldPath.Child("secretRef", "namespace"))...)
+		}
+	}
```
Please create a common function for `SecretRef` validation instead of copying it around.

akshitgrover replied:
I was actually thinking of doing it.
I had one question though, Should this be inside the apimachinery package? As SecretReference object is nothing specific to volumes?

jsafrane replied:
I think it should be here, in validation.go, because `SecretReference` is declared in `pkg/apis/core` too

akshitgrover replied:
Added [ValidateSecretReference](https://github.com/kubernetes/kubernetes/pull/103024/files#diff-c713e8919642d873fdf48fe8fb6d43e5cb2f53fd601066ff53580ea655948f0dR292) function

jsafrane commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1262,6 +1271,21 @@ func validateRBDPersistentVolumeSource(rbd *core.RBDPersistentVolumeSource, fldP
+		if len(rbd.SecretRef.Namespace) == 0 {
+			allErrs = append(allErrs, field.Required(fldPath.Child("secretRef", "namespace"), ""))
+		} else {
+			allErrs = append(allErrs, ValidateDNS1123Label(rbd.SecretRef.Namespace, fldPath.Child("secretRef", "namespace"))...)
```
It could be useful to declare `ValidateNamespaceName`

akshitgrover replied:
This makes sense, Will add a function for the same

akshitgrover replied:
ValidateNamespaceName function already exists.
[ValidateDNS1123Label](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L114) does the same thing as that of ValidateNamespaceName the only difference the return signature.
[ValidateDNS1123Label](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/validation/validation.go#L114) return field.ErrorList and ValidateNamespaceName returns []string using which field.ErrorList has to be created.

jsafrane commented:
/ok-to-test

akgrover committed:
Resolve comments

- Add ValidateSecretReference func

Signed-off-by: Akshit Grover <akshit.grover2016@gmail.com>

akshitgrover commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1507,12 +1539,10 @@ func validateStorageOSPersistentVolumeSource(storageos *core.StorageOSPersistent
-		if len(storageos.SecretRef.Namespace) == 0 {
-			allErrs = append(allErrs, field.Required(fldPath.Child("secretRef", "namespace"), ""))
-		}
+		allErrs = append(allErrs, ValidateSecretReference(&core.SecretReference{
```
I created a new core.SecreteReference object since storageos.SecretReference is of ObjectReference.
I am curious why is it declared as type ObjectReference.

akshitgrover commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1507,12 +1539,10 @@ func validateStorageOSPersistentVolumeSource(storageos *core.StorageOSPersistent
-		if len(storageos.SecretRef.Namespace) == 0 {
-			allErrs = append(allErrs, field.Required(fldPath.Child("secretRef", "namespace"), ""))
-		}
+		allErrs = append(allErrs, ValidateSecretReference(&core.SecretReference{
+			Name: storageos.SecretRef.Name,
+			Namespace: storageos.SecretRef.Namespace,
+		}, fldPath.Child("secretRef"))...)
```
I created a new core.SecreteReference object since storageos.SecretReference is of ObjectReference.
I am curious why is it declared as type ObjectReference.

akshitgrover replied:
Updated this as ValidateSecretObjectReference method is added for the same

jsafrane commented on pkg/apis/core/validation/validation.go:
```diff
@@ -286,6 +286,26 @@ var ValidateClassName = apimachineryvalidation.NameIsDNSSubdomain
+// ValidateSecretReference can be use to check whether provided SecretReference object is valid.
+// Secret name is provided and is valid
+// Secret namespace is provided and is valid
+func ValidateSecretReference(secretRef *core.SecretReference, fldPath *field.Path) field.ErrorList {
```
Please add unit tests for this function.

akshitgrover replied:
[TestValidateSecretReference](https://github.com/kubernetes/kubernetes/pull/103024/files#diff-7ee023b4401073292170227985c8ac98015db0c1980657c277a7cea011c24830R17839) function added

akgrover committed:
Resolve comments

- Add unit tests
- Fix formatting errors

Signed-off-by: Akshit Grover <akshit.grover2016@gmail.com>

jsafrane commented on pkg/apis/core/validation/validation.go:
```diff
@@ -1507,12 +1539,10 @@ func validateStorageOSPersistentVolumeSource(storageos *core.StorageOSPersistent
-		if len(storageos.SecretRef.Namespace) == 0 {
-			allErrs = append(allErrs, field.Required(fldPath.Child("secretRef", "namespace"), ""))
-		}
+		allErrs = append(allErrs, ValidateSecretReference(&core.SecretReference{
```
I would prefer if there was`ValidateLocalSecretReference()` instead of calling ValidateSecretReference with a fake object.

akshitgrover replied:
I understand the concern but I don't quite get the solution
At the risk of sounding lame I have to ask 
Can you please share some reference?

jsafrane replied:
Sorry, `ValidateLocalSecretReference` was not a good name. `storageos.SecretRef` is a pointer to `ObjectReference`. I am proposing:

`func ValidateSecretObjectReference(ref *core.ObjectReference) []error` that checks if `ObjectReference` is really a Secret + has valid name + namespace.

akshitgrover replied:
This makes sense
I'll update the PR

akshitgrover replied:
@jsafrane 
I have added ValidateSecretObjectReference method that follows the criteria you mentioned above

fejta commented:
/uncc
Unsure why the bot added me to this PR

akgrover committed:
Add ValidateSecretObjectReference method

- Add ValidateSecretObjectReference method to validate StorageOSPersistentVolumeSource.SecretRef
- Add unit tests for the same method

Signed-off-by: Akshit Grover <akshit.grover2016@gmail.com>

jsafrane commented on pkg/apis/core/validation/validation.go:
```diff
@@ -286,6 +286,47 @@ var ValidateClassName = apimachineryvalidation.NameIsDNSSubdomain
+func ValidateSecretObjectReference(secretRef *core.ObjectReference, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if secretRef.Kind != "Secret" {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("kind"), secretRef.Kind, "kind should be a secret"))
```
`kind should be Secret`

akshitgrover replied:
Fixed the description

jsafrane commented:
/assign @liggitt
for approval. Not sure if it is an API change - it may render some API objects not valid, however, those objects are unusable by Kubernetes anyway, as it can't get the secrets with invalid names. Now the user gets an error earlier.

akgrover committed:
Fix error description

Signed-off-by: Akshit Grover <akshit.grover2016@gmail.com>

akshitgrover commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

akshitgrover commented:
/retest

liggitt commented:
/hold

standard questions when we want to tighten validation:
* does it prevent a no-op update of existing persisted objects that would fail this new validation? (would block storage migration tasks)
* does it prevent deletion of existing persisted objects with finalizers that would fail this new validation? (update is required to remove finalizers to allow deletion)
* does it prevent controllers from updating status of existing persisted objects that would fail this new validation?

we could consider ratcheting validation:
* apply this tightened validation on create
* apply this tightened validation on update where the old object passed this validation (don't let a valid object become invalid)
* skip this tightened validation on update where the old object fails this validation

c.f. https://github.com/kubernetes/kubernetes/issues/59984

liggitt commented:
> it may render some API objects not valid, however, those objects are unusable by Kubernetes anyway, as it can't get the secrets with invalid names. Now the user gets an error earlier.

It would be good to note the places where an empty secret names produces a fatal error 100% of the time. Those are the only places we would consider tightening validation, but even there, we would not simply start marking the object as invalid... the validation would need to be aware of whether it was validating an update of an existing object that already contained empty secret name references, and tolerate the update in that case.

I wouldn't recommend refactoring existing validation AND adding tightening validation in the same PR... it makes it really hard to review to be sure we're not breaking compatibility.

akshitgrover commented:
@jsafrane 
Can you help me answer some of these questions?
Thank you

jsafrane commented:
I think Jordan is right in all points, this PR needs to be reworked to the ratcheting validation he suggests. Otherwise people won't be able to delete their newly invalid objects from the API server.

akshitgrover commented:
@jsafrane 
is there any reference from which I can understand what should actually be implemented
I kind of got lost there in the words
Sorry for being lame :')

akshitgrover commented:
@jsafrane 
I was going through the code the other and I might have figured out a way to implement ratcheted validation
I'll raise a PR with the update

akshitgrover commented:
@jsafrane 
Does it make sense to update this function to accept olvPVSpec object as a parameter and same thing will be done for all volume specific validation functions
We'll compare the old spec with new spec if there is any difference then we'll run the newly added secretRef validation otherwise we'll skip it, In case of create we can keep the old object as null (This approach covers all the points mentioned by @liggitt for ratcheted validation)

Just needed to get your view on this before I start with the implementation

akgrover committed:
Update validation code to be backward compatible

New tightened validation should be applied only in the case of:
1. New object is created
2. Field which is to be validated differs in new object from old object

Signed-off-by: Akshit Grover <akshit.grover2016@gmail.com>

akgrover committed:
Update gofmt

Signed-off-by: Akshit Grover <akshit.grover2016@gmail.com>

akshitgrover commented:
@jsafrane 
I have updated the PR with the ratcheted validation 
Newly added validation runs only on 2 scenarios
1. Creation of new object
2. Object field which is to be validated differs in the updated object than old object

I am yet to add unit tests for these scenarios, PTAL at the implementation
Thank you

akshitgrover commented:
/test pull-kubernetes-integration

akshitgrover commented:
/test pull-kubernetes-integration

liggitt commented:
/assign @roycaihw

roycaihw reviewed (commented):
Thanks for opening the PR! It's clearly an advanced work for a new contributor, and I think it's in the right direction :)

I think it would be safer and easier to review if we can separate the work into several smaller PRs:
- relaxing too-tight validation
- refactoring
- tightening too-relax validation

akshitgrover commented on pkg/apis/core/validation/validation.go:
```diff
@@ -286,6 +286,47 @@ var ValidateClassName = apimachineryvalidation.NameIsDNSSubdomain
+	if len(secretRef.Name) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("name"), ""))
+	} else {
+		for _, msg := range ValidateSecretName(secretRef.Name, false) {
```
This makes sense
I'll raise a PR with the change
Thank you for the review

akshitgrover commented on pkg/apis/core/validation/validation.go:
```diff
@@ -806,10 +847,8 @@ func validateISCSIPersistentVolumeSource(iscsi *core.ISCSIPersistentVolumeSource
-	if iscsi.SecretRef != nil {
-		if len(iscsi.SecretRef.Name) == 0 {
-			allErrs = append(allErrs, field.Required(fldPath.Child("secretRef", "name"), ""))
-		}
```
I think I might be missing something here
When we say SecretReference don't we mean both namespace and name, I mean if either of these is invalid then the secret itself is un-usable and thus the respective volume controller will not be able to use it anyway
cc @jsafrane

roycaihw replied:
I agree a secret cannot be referred to if the name/namespace is missing from the secret reference, but the partial secret reference doesn't necessarily break the respective volume controller. 

The secret reference is an optional field in some PersistentVolumeSources. I think the first concern discussed in https://github.com/kubernetes/kubernetes/pull/103024#issuecomment-874848298 was if a PersistentVolumeSource tolerates a partial secret reference (e.g. treats it as nil) and doesn't return a fatal error 100% of the time, then we cannot tighten the validation; otherwise we may break existing user workflow when we upgrade

dims commented:
this PR has been on hold for about 2 weeks. Can you please see how to unblock this by reviewing comments etc? or please close this out (Trying to declutter our open list of PRs)

dims commented:
Looks like this needs a rebase too @akshitgrover

dims commented:
This PR is older than 4 weeks and needs as rebase.

Please rebase the PR against latest master and reopen if still needed.

/close
